rust/capture/src/passes/invalid_surrogates.rs

```diff
@@ -0,0 +1,226 @@
+use std::str::Chars;
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+enum LastSeen {
+    Char,   // Any regular character
+    Escape, // We've seen a backslash, not preceded by another backslash
+}
+
+// Unicode unknown character replacement - �, but as a hex escape sequence
+const REPLACEMENT: &str = "uFFFD";
+const HIGH_SURROGATE_RANGE: std::ops::Range<u16> = 0xD800..0xDBFF;
+const LOW_SURROGATE_RANGE: std::ops::Range<u16> = 0xDC00..0xDFFF;
+
+pub struct InvalidSurrogatesPass<'a> {
+    input: Chars<'a>,
+    last_seen: LastSeen,
+    pending_output: Vec<char>,
+    pending_ptr: usize,
+    escape_seq_buf: String,
+}
+
+impl<'a> Iterator for InvalidSurrogatesPass<'a> {
+    type Item = char;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.step()
+    }
+}
+
+impl<'a> InvalidSurrogatesPass<'a> {
+    // This is a simple heuristic to determine if we need to run this pass
+    pub fn needed(input: &str) -> bool {
+        input.contains("\\u")
```

Comment:
I know it's simple, but 2 tiny test cases to verify true/false where expected would be nice... this is the kind of thing someone will change on a whim one day and break everything without knowing.

Reply:
Removed instead, premature opt